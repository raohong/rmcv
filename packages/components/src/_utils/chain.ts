const chain = <T extends (...args: any[]) => void>(
  ...fns: (T | undefined)[]
): ((...args: Parameters<T>) => void) => {
  return (...args) => {
    fns.forEach((item) => {
      if (item) {
        item(...args);
      }
    });
  };
};

export default chain;
