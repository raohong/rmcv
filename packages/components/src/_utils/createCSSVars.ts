import kebabCase from 'lodash/kebabCase';

const createCSSVars = (
  prefix: string,
  style: Record<string, string | undefined | null>,
) => {
  const result = Object.entries(style).reduce((map, [key, value]) => {
    if (!value) {
      return map;
    }
    // @ts-ignore
    // eslint-disable-next-line no-param-reassign
    map[`--${prefix}-${kebabCase(key)}`] = value;

    return map;
  }, {} as React.CSSProperties);

  return result;
};

export default createCSSVars;
