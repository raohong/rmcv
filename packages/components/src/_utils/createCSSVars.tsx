import React from 'react';

export const createCSSVars = (vars: Record<string, number | string>) =>
  Object.entries(vars).reduce((result, [name, value]) => {
    // @ts-ignore
    result[`--${name}`] = value;
    return result;
  }, {} as React.CSSProperties);
