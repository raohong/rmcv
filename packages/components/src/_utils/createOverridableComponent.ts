import React, { RefAttributes } from 'react';

interface OverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
> extends React.FC<
    Omit<
      Omit<React.ComponentProps<C>, 'ref'> &
        Omit<
          JSX.IntrinsicElements[DefaultElement],
          Exclude<'ref', keyof React.ComponentProps<C>>
        >,
      keyof ExcludeProps | 'component'
    >
  > {
  // JSX.IntrinsicElements
  <Tag extends keyof JSX.IntrinsicElements>(
    props: Omit<
      Omit<
        JSX.IntrinsicElements[Tag],
        Exclude<'ref', keyof React.ComponentProps<C>>
      > &
        Omit<React.ComponentProps<C>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Tag;
    },
  ): JSX.Element;

  // ClassComponent
  <Custom extends React.ComponentClass<React.ComponentProps<Custom>>>(
    props: Omit<
      Omit<React.ComponentProps<C> & React.ComponentProps<Custom>, 'ref'>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    } & RefAttributes<InstanceType<Custom>>,
  ): JSX.Element;

  <
    Custom extends React.ForwardRefExoticComponent<
      React.ComponentProps<Custom>
    >,
  >(
    props: Omit<
      Omit<React.ComponentProps<C>, 'ref'> & React.ComponentProps<Custom>,
      keyof ExcludeProps | 'component'
    > & {
      component: Custom;
    },
  ): JSX.Element;
}

function createOverridableComponent<
  C extends React.ComponentType<React.ComponentProps<C>>,
  DefaultElement extends keyof JSX.IntrinsicElements = 'div',
  ExcludeProps extends object = object,
>(component: C): OverridableComponent<C, DefaultElement, ExcludeProps> {
  return component as unknown as OverridableComponent<
    C,
    DefaultElement,
    ExcludeProps
  >;
}

export default createOverridableComponent;
