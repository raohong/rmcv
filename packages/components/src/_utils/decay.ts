import { useEffect, useRef, useCallback } from 'react';
import type { SpringValue } from '@react-spring/web';
import raf from 'raf';

export type DecayAnimationOptions = {
  power?: number;
  timeConst?: number;
  from?: number;
  velocity?: number;
  modifyTarget?: (dist: number, velocity: number) => number;
  restDelta?: number;
};

type DecayAnimationState = {
  finished: boolean;
  value: number;
  velocity: number;
};

const decay = ({
  power = 0.8,
  timeConst = 300,
  from = 0,
  velocity = 0,
  restDelta = 0.2,
  modifyTarget,
}: DecayAnimationOptions) => {
  let movement = velocity * power * timeConst;
  const ideal = movement + from;

  const target = modifyTarget ? modifyTarget(ideal, velocity) : ideal;
  const state: DecayAnimationState = {
    finished: false,
    value: from,
    velocity,
  };

  if (target !== ideal) {
    movement = target - from;
  }

  return {
    next: (t: number) => {
      // y = a * Math.exp(kt)

      const dx = Math.exp((1 / -timeConst) * t);
      const delta = -movement * dx;

      state.finished = Math.abs(delta) < restDelta;
      state.value = state.finished ? target : target + delta;
      state.velocity = velocity * dx;

      return state;
    },
  };
};

export const runDecayAnimation = ({
  onChange,
  ...options
}: DecayAnimationOptions & {
  onChange: (state: DecayAnimationState) => void;
}) => {
  let id: number | null = null;
  let active = true;
  const startTime = Date.now();

  const generator = decay(options);

  const cancel = () => {
    if (id !== null) {
      raf.cancel(id);
      id = null;
    }

    active = false;
  };

  const run = () => {
    id = raf(() => {
      if (active) {
        const now = Date.now();
        const state = generator.next(now - startTime);

        onChange(state);

        if (!state.finished) {
          run();
        }
      }
    });
  };

  run();

  return cancel;
};

export const useDecayAnimation = () => {
  const unmountedRef = useRef(false);
  const canceller = useRef<(() => void)[]>([]);

  const cancelAnimation = useCallback(() => {
    if (canceller.current) {
      canceller.current.forEach((item) => item());
    }
  }, []);

  const runDecay = useCallback(
    (
      spring: SpringValue<number>,
      dir: number,
      boundary: number,
      options: DecayAnimationOptions,
    ) => {
      const cb = runDecayAnimation({
        ...options,
        onChange({ velocity, value }) {
          if (unmountedRef.current) {
            return;
          }

          if (
            (dir === 1 && value >= boundary) ||
            (dir === -1 && value <= boundary)
          ) {
            cb();
            canceller.current = canceller.current.filter((item) => item !== cb);

            spring.start({
              to: boundary,
              config: {
                decay: false,
                velocity,
              },
            });
          } else {
            spring.set(value);
          }
        },
      });

      canceller.current.push(cb);
    },
    [],
  );

  useEffect(() => {
    return () => {
      cancelAnimation();
    };
  }, [cancelAnimation]);

  return { runDecay, cancelAnimation };
};
