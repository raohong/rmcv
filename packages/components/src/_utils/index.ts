export { default as getPrefixCls } from './getPrefixCls';

export { default as isBrowser } from './isBrowser';

export { default as renderPortal } from './renderPortal';

export { default as uuid } from './uuid';

export { default as createCSSVars } from './createCSSVars';

export { default as chain } from './chain';

export { default as toArray } from './toArray';
