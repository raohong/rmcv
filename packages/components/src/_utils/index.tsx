export * from './createOverridableComponent';
export * from './decay';
export * from './flatReactNode';
export * from './getPrefixCls';
export * from './htmlAttrs';
export * from './reactRender';
export * from './useXId';
