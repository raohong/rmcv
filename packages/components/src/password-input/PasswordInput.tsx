import { useIsomorphicLayoutEffect } from '@rmc-vant/hooks';
import { isEmpty, isNumber, omit } from '@rmc-vant/utils';
import classNames from 'classnames';
import React, { useEffect, useMemo, useRef, useState } from 'react';
import { useConfigContext } from '../config-provider';
import type { PasswordInputProps } from './interface';

const getItemList = (length: number) => {
  return Array.from(
    { length: isNumber(length) ? Math.max(0, length) : 0 },
    (_, i) => i,
  );
};

const PasswordInput = React.forwardRef<HTMLLabelElement, PasswordInputProps>(
  (
    {
      onFocus,
      className,
      errorInfo,
      info,
      gutter,
      value,
      mask = true,
      length = 6,
      ...props
    },
    ref,
  ) => {
    const has = 'focused' in props;
    const focusedFromProps = !!props.focused;
    const innerDOMRef = useRef<HTMLDivElement>(null);

    const { getPrefixCls } = useConfigContext();
    const cls = getPrefixCls('password-input');
    const [focused, setFocused] = useState(has ? focusedFromProps : false);
    const [hasGutter, setHasGutter] = useState(() => {
      if (isEmpty(gutter)) {
        return true;
      }

      const target = parseFloat(String(gutter));

      return Number.isNaN(target) ? true : target > 0;
    });

    const internalFocused = has ? focusedFromProps : focused;

    const handleFocus = () => {
      onFocus?.();

      if (!has) {
        setFocused(true);
      }
    };

    const handleBlur = () => {
      if (!has) {
        setFocused(false);
      }
    };

    useEffect(() => {
      if (!has) {
        setFocused(focusedFromProps);
      }
    }, [has, focusedFromProps]);

    useIsomorphicLayoutEffect(() => {
      const el = innerDOMRef.current;
      if (!el) {
        return;
      }

      const gap = parseFloat(getComputedStyle(el).columnGap);

      setHasGutter(gap > 0);
    }, [gutter, innerDOMRef]);

    const itemList = useMemo(() => getItemList(length), [length]);
    const internalValue = isEmpty(value) ? '' : String(value).slice(0, length);
    const currentValueLength = internalValue.length;
    const valueList = internalValue.split('');

    return (
      <label
        className={classNames(
          cls,
          { [`${cls}-focused`]: internalFocused },
          className,
        )}
        onFocus={handleFocus}
        onBlur={handleBlur}
        ref={ref}
        {...omit(props, ['focused'])}
      >
        <div
          style={{
            columnGap: gutter,
          }}
          ref={innerDOMRef}
          className={classNames(`${cls}-inner`, !hasGutter && `${cls}-inner-inset`)}
        >
          {itemList.map((item, index) => {
            let content: React.ReactNode = valueList[item];

            if (index === currentValueLength && internalFocused) {
              content = <span className={`${cls}-cursor`} />;
            } else if (mask && index < currentValueLength) {
              content = <span className={`${cls}-mask-dot`} />;
            }

            return (
              <div key={item} className={`${cls}-item`}>
                {content}
              </div>
            );
          })}
          <input
            value={internalValue}
            maxLength={length}
            type="password"
            className={`${cls}-input`}
          />
        </div>
        <div
          className={classNames(
            `${cls}-info`,
            !isEmpty(errorInfo) && `${cls}-error-info`,
          )}
        >
          {isEmpty(errorInfo) ? info : errorInfo}
        </div>
      </label>
    );
  },
);

export default PasswordInput;
