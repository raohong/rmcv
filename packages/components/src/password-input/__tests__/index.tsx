import React from 'react';
import { render, screen } from '@testing-library/react';
import userEvent from '@testing-library/user-event';
import PasswordInput from '../PasswordInput';
import { getPrefixCls } from '../../_utils';

const testId = 'password-input';

test('render correctly', () => {
  const tree = render(<PasswordInput />);

  expect(tree.asFragment()).toMatchSnapshot();
});

test('render with length', () => {
  render(<PasswordInput data-testid={testId} length={6} />);

  expect(
    screen
      .getByTestId(testId)
      .querySelectorAll(`.${getPrefixCls('password-input')}-item`).length,
  ).toBe(6);
});

test('render with value', () => {
  render(<PasswordInput data-testid={testId} value="123" />);

  expect(
    screen
      .getByTestId(testId)
      .querySelector(`.${getPrefixCls('password-input-input')}`),
  ).toHaveValue('123');
});

test('render with gutter', () => {
  render(<PasswordInput data-testid={testId} gutter={10} />);

  const inner = screen
    .getByTestId(testId)
    .querySelector(`.${getPrefixCls('password-input-inner')}`);

  expect(inner).toHaveStyle({
    'column-gap': '10px',
  });
  expect(inner).not.toHaveClass(getPrefixCls('password-input-inner-inset'));
});

test('render with mask', () => {
  render(<PasswordInput data-testid={testId} value="12" />);

  expect(
    screen
      .getByTestId(testId)
      .querySelectorAll(`.${getPrefixCls('password-input-mask-dot')}`).length,
  ).toBe(2);
});

test('render with focused', () => {
  render(<PasswordInput data-testid={testId} focused />);

  expect(
    screen
      .getByTestId(testId)
      .querySelector(`.${getPrefixCls('password-input-cursor')}`),
  ).toBeInTheDocument();
});
