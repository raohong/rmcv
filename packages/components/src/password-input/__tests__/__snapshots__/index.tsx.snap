// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render correctly 1`] = `
<DocumentFragment>
  <label
    class="rmcv-password-input"
  >
    <div
      class="rmcv-password-input-inner rmcv-password-input-inner-inset"
    >
      <div
        class="rmcv-password-input-item"
      />
      <div
        class="rmcv-password-input-item"
      />
      <div
        class="rmcv-password-input-item"
      />
      <div
        class="rmcv-password-input-item"
      />
      <div
        class="rmcv-password-input-item"
      />
      <div
        class="rmcv-password-input-item"
      />
      <input
        class="rmcv-password-input-input"
        maxlength="6"
        type="password"
        value=""
      />
    </div>
    <div
      class="rmcv-password-input-info"
    />
  </label>
</DocumentFragment>
`;
